gcode.c

```diff
@@ -80,368 +78,611 @@ static uint8_t gc_check_same_position(float *pos_a, float *pos_b)
   }
   return(true);
 }
-         
-// Executes one line of 0-terminated G-Code. The line is assumed to contain only uppercase
-// characters and signed floating point values (no whitespace). Comments and block delete
-// characters have been removed. In this function, all units and positions are converted and 
-// exported to grbl's internal functions in terms of (mm, mm/min) and absolute machine 
-// coordinates, respectively.
-uint8_t gc_execute_line(char *line) 
+
+bool gc_is_value_in_array(uint8_t val, uint8_t arr[], uint8_t len)
 {
-  /* -------------------------------------------------------------------------------------
-     STEP 1: Initialize parser block struct and copy current g-code state modes. The parser
-     updates these modes and commands as the block line is parser and will only be used and
-     executed after successful error-checking. The parser block struct also contains a block
-     values struct, word tracking variables, and a non-modal commands tracker for the new 
-     block. This struct contains all of the necessary information to execute the block. */
-     
-  memset(&gc_block, 0, sizeof(gc_block)); // Initialize the parser block struct.
-  memcpy(&gc_block.modal,&gc_state.modal,sizeof(gc_modal_t)); // Copy current modes
-  uint8_t axis_command = AXIS_COMMAND_NONE;
-  uint8_t axis_0, axis_1, axis_linear;
-  float coordinate_data[N_AXIS]; // Multi-use variable to store coordinate data for execution
-  float parameter_data[N_AXIS]; // Multi-use variable to store parameter data for execution
+  uint8_t idx;
+  for (idx = 0; idx < len; idx++) {
+    if (arr[idx] == val)
+      return true;
+  }
+  return false;
+
+}
+
+uint8_t gc_process_modal_group_G0(uint8_t int_value, uint8_t mantissa, uint8_t * axis_command)
+{
+  (void)axis_command;
+  switch(int_value) {
+    case 4: 
+      gc_block.non_modal_command = NON_MODAL_DWELL;
+      break; // G4
+    case 10: 
+      gc_block.non_modal_command = NON_MODAL_SET_COORDINATE_DATA;
+      break; // G10
+    case 28:
+      switch(mantissa) {
+        case 0: 
+          gc_block.non_modal_command = NON_MODAL_GO_HOME_0; 
+          break;  // G28
+        case 10:
+          gc_block.non_modal_command = NON_MODAL_SET_HOME_0;
+          break; // G28.1
+        default:
+          FAIL(STATUS_GCODE_UNSUPPORTED_COMMAND); // [Unsupported G28.x command]
+      }
+      mantissa = 0; // Set to zero to indicate valid non-integer G command.
+      break;
+    case 30: 
+      switch(mantissa) {
+        case 0: gc_block.non_modal_command = NON_MODAL_GO_HOME_1; break;  // G30
+        case 10: gc_block.non_modal_command = NON_MODAL_SET_HOME_1; break; // G30.1
```

Comment:
Since this is a major refactor, can we stick to a single style of case statement and not do single line `case: statement; break;`